### PR TITLE
Tally reaction rates on mesh -- 3D MOC

### PR DESCRIPTION
**Pull request by tjlaboss**
PR #380 , but for 3D-MOC.

Needs to include the 'absorption' reaction.

`Material.getSigmaAByGroup()` isn't in the C++ source code. I'll see what needs to be done to add it, and then include it in the unit test.

### CODE REVIEW

File: openmoc/process.py [hunk 1]
```diff
@@ -213,6 +215,48 @@ def compute_fission_rates(solver, use_hdf5=False):
...
+        For scatter, this includes self-scatter and outscatter.
+
+    """
+    global rxn_types
+    cv.check_value('rxn_type', rxn_type, rxn_types)
+    
+    # Energy groups start at 1 in OpenMOC
+    if rxn_type == "total":
+        return material.getSigmaTByGroup(g + 1)
+    elif rxn_type == "absorption":
```
Comment (GiudGiud) [inline openmoc/process.py]: remove absorption unless you add a "getSigmaAByGroup"

Comment (tjlaboss) [reply openmoc/process.py]: done

**Comment by tjlaboss** [comment]
I did some digging. I found that `Mesh.tally_fission_rates(...)` finds "fission" rates in develop, but "nu-fission" rates in 3D-MOC. Is this intentional?

I will update `test_mesh_tally` accordingly.

**Comment by GiudGiud** [comment]
Not sure if it's intentional.
It's somewhat of a bug for now, if you want to fix it now, add a boolean to choose between fission and nu-fission rates in solver.computeFSRFissionRates(). Default should be fission

File: src/CPUSolver.cpp [hunk 2]
```diff
@@ -2202,44 +2202,50 @@ void CPUSolver::stabilizeFlux() {
...
   /* Initialize fission rates to zero */
   for (long r=0; r < _num_FSRs; r++)
     fission_rates[r] = 0.0;
 
   /* Loop over all FSRs and compute the volume-weighted fission rate */
-#pragma omp parallel for private (nu_sigma_f, vol) schedule(static)
+#pragma omp parallel for private (sigma_f, vol) schedule(static)
   for (long r=0; r < _num_FSRs; r++) {
-    nu_sigma_f = _FSR_materials[r]->getNuSigmaF();
+    /* TODO: move the 'if' block outside */
```
Comment (tjlaboss) [inline src/CPUSolver.cpp]: @GiudGiud, I don't have much experience with C++. Is it possible to move the `if-else` block before the `for` loop and select the getSigmaF() function there?

File: src/CPUSolver.cpp [hunk 3]
```diff
@@ -2202,44 +2202,50 @@ void CPUSolver::stabilizeFlux() {
...
  * @endcode
  *
- * @param fission_rates an array to store the nu-fission rates (implicitly
+ * @param fission_rates an array to store the fission rates (implicitly
  *                      passed in as a NumPy array from Python)
  * @param num_FSRs the number of FSRs passed in from Python
+ * @param nu whether to use nu-fission rates instead of fission rates
  */
-void CPUSolver::computeFSRFissionRates(double* fission_rates, long num_FSRs) {
+void CPUSolver::computeFSRFissionRates(double* fission_rates, long num_FSRs, bool nu = false) {
```
Comment (GiudGiud) [inline src/CPUSolver.cpp]: line too long

File: src/CPUSolver.cpp (hunk 2 above)
Comment (GiudGiud) [reply src/CPUSolver.cpp]: It's fine like that. The penalty for an if statement here is very small

File: src/CPUSolver.cpp [hunk 4]
```diff
@@ -2202,44 +2202,50 @@ void CPUSolver::stabilizeFlux() {
...
 
   /* Loop over all FSRs and compute the volume-weighted fission rate */
-#pragma omp parallel for private (nu_sigma_f, vol) schedule(static)
+#pragma omp parallel for private (sigma_f, vol) schedule(static)
   for (long r=0; r < _num_FSRs; r++) {
-    nu_sigma_f = _FSR_materials[r]->getNuSigmaF();
+    /* TODO: move the 'if' block outside */
+    if (nu) {
+      sigma_f = _FSR_materials[r]->getNuSigmaF();
+    } else {
```
Comment (GiudGiud) [inline src/CPUSolver.cpp]: }
else {

File: src/CPUSolver.h [hunk 5]
```diff
@@ -123,7 +123,7 @@ class CPUSolver : public Solver {
   int getNumThreads();
   void setNumThreads(int num_threads);
   void setFixedSourceByFSR(long fsr_id, int group, FP_PRECISION source);
-  void computeFSRFissionRates(double* fission_rates, long num_FSRs);
+  void computeFSRFissionRates(double* fission_rates, long num_FSRs, bool nu);
```
Comment (GiudGiud) [inline src/CPUSolver.h]: bool nu = false in the header is better

File: src/Solver.cpp [hunk 6]
```diff
@@ -1747,7 +1747,7 @@ void Solver::printFissionRates(std::string fname, int nx, int ny, int nz) {
 
   int num_fsrs = _geometry->getNumTotalFSRs();
   double* fsr_fission_rates = new double[num_fsrs];
-  computeFSRFissionRates(fsr_fission_rates, num_fsrs);
+  computeFSRFissionRates(fsr_fission_rates, num_fsrs, false);
```
Comment (GiudGiud) [inline src/Solver.cpp]: no need to specify if it's the default

File: src/CPUSolver.cpp [hunk 7]
```diff
@@ -2202,44 +2202,50 @@ void CPUSolver::stabilizeFlux() {
...
  * @code
  *          num_FSRs = geometry.getNumFSRs()
  *          fission_rates = solver.computeFSRFissionRates(num_FSRs)
  * @endcode
  *
- * @param fission_rates an array to store the nu-fission rates (implicitly
+ * @param fission_rates an array to store the fission rates (implicitly
  *                      passed in as a NumPy array from Python)
  * @param num_FSRs the number of FSRs passed in from Python
+ * @param nu whether to use nu-fission rates instead of fission rates
```
Comment (GiudGiud) [inline src/CPUSolver.cpp]: * @param nu whether return nu-fission (true) or fission (default, false) rates

**Comment by tjlaboss** [comment]
Thanks for the review, Guillaume. I've applied the changes in the latest commit.

**Pull request merged**